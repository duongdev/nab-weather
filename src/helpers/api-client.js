import axios from 'axios'

const client = axios.create({
  baseURL:
    process.env.NODE_ENV === 'development'
      ? undefined
      : 'https://cors-anywhere.herokuapp.com/https://www.metaweather.com/api',
})

export const fetchLocations = async (searchText = '') => {
  if (!searchText) {
    return []
  }
  const { data } = await client.get(`/location/search/`, {
    params: { query: searchText },
  })

  return data
}

export const fetchLocationWeather = async (locationId) => {
  const { data } = await client.get(`/location/${locationId}/`)

  return data
}

export default client
