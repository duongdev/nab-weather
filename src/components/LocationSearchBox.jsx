import React, { useState, useEffect, useCallback } from 'react'
import { Autocomplete } from '@material-ui/lab'
import { TextField } from '@material-ui/core'
import { useQuery } from 'react-query'
import parse from 'autosuggest-highlight/parse'
import match from 'autosuggest-highlight/match'
import { fetchLocations } from 'helpers/api-client'

const LocationSearchBox = (props) => {
  const [inputValue, setInputValue] = useState('')
  const [value, setValue] = useState(null)
  const [options, setOptions] = useState([])
  const { data, isLoading, error, isError } = useQuery(
    ['locations', inputValue.trim()],
    () => fetchLocations(inputValue.trim()),
    {
      // initialData: [],
      enabled: inputValue,
    },
  )

  const handleChangeLocation = useCallback(
    (event, location) => {
      setValue(location)
      // eslint-disable-next-line no-unused-expressions
      props.onChange?.(location)
    },
    [props.onChange],
  )

  useEffect(() => {
    if (!inputValue) return setOptions(value ? [value] : [])
    if (data) return setOptions(data)
  }, [data, inputValue, value])

  return (
    <Autocomplete
      autoComplete
      includeInputInList
      id="location-search"
      options={options ?? []}
      getOptionLabel={(location) => location.title}
      filterOptions={(x) => x}
      onInputChange={(event, value) => {
        setInputValue(value)
      }}
      onChange={handleChangeLocation}
      value={value}
      loading={isLoading}
      inputValue={inputValue}
      getOptionSelected={(opt, val) => opt?.woeid === val?.woeid}
      renderInput={(params) => (
        <TextField
          {...params}
          autoFocus
          fullWidth
          placeholder="Search a location..."
          variant="outlined"
          color={isError ? 'error' : 'primary'}
          helperText={error && error.message}
        />
      )}
      renderOption={(option, { inputValue }) => {
        const matches = match(option.title, inputValue)
        const parts = parse(option.title, matches)

        return (
          <div data-testid={`option-${option.woeid}`}>
            {parts.map((part, index) => (
              <span
                key={index}
                style={{ fontWeight: part.highlight ? 700 : 400 }}
              >
                {part.text}
              </span>
            ))}
          </div>
        )
      }}
    />
  )
}

export default LocationSearchBox
