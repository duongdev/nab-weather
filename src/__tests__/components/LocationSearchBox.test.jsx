import React from 'react'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import LocationSearchBox from 'components/LocationSearchBox'
import * as api from 'helpers/api-client'

describe('<LocationSearchBox />', () => {
  it(`renders a text box`, () => {
    render(<LocationSearchBox />)
    expect(
      screen.getByPlaceholderText('Search a location...'),
    ).toBeInTheDocument()
    expect(screen.getByRole('textbox')).toBeInTheDocument()
  })

  it(`calls API to search for locations and calls onChange`, async () => {
    const fetchLocationsFn = jest
      .spyOn(api, 'fetchLocations')
      .mockResolvedValueOnce(
        Promise.resolve([
          {
            title: 'Ho Chi Minh City',
            location_type: 'City',
            woeid: 1252431,
            latt_long: '10.759180,106.662498',
          },
        ]),
      )
    const onChange = jest.fn()

    render(<LocationSearchBox onChange={onChange} />)
    const input = screen.getByRole('textbox')
    fireEvent.change(input, { target: { value: 'Ho Chi Minh' } })
    await waitFor(() => {
      expect(screen.getByRole('option').innerHTML).toMatchSnapshot()
      expect(fetchLocationsFn).toBeCalledWith('Ho Chi Minh')
      const hcmOption = screen.getByTestId(`option-1252431`)
      fireEvent.click(hcmOption)
      expect(onChange).toBeCalledWith({
        title: 'Ho Chi Minh City',
        location_type: 'City',
        woeid: 1252431,
        latt_long: '10.759180,106.662498',
      })
    })
  })
})
