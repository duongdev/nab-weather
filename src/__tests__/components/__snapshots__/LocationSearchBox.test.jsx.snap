// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationSearchBox /> calls API to search for locations and calls onChange 1`] = `"<div data-testid=\\"option-1252431\\"><span style=\\"font-weight: 700;\\">Ho</span><span style=\\"font-weight: 400;\\"> </span><span style=\\"font-weight: 700;\\">Chi</span><span style=\\"font-weight: 400;\\"> </span><span style=\\"font-weight: 700;\\">Minh</span><span style=\\"font-weight: 400;\\"> City</span></div>"`;
